### PR TITLE
feat: Custom attributes in automations and refactor

### LINKED ISSUE #4458: Add Custom Attributes support in automation
- Update the UI to support custom attributes for conversation / contact.

### PR DESCRIPTION
**Pull request by fayazara**
## Description

Adds custom attributes to automations and refactoring of automations module.

Fixes #4458 

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
Test specs for mixins and helpers

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by pranavrajs** [comment]
@fayazara Is it ready to be reviewed?

File: app/javascript/dashboard/routes/dashboard/settings/automation/AddAutomationRule.vue [hunk 1]
```diff
@@ -259,7 +260,35 @@ export default {
...
+      return {
+        key: attr.attribute_key,
+        name: attr.attribute_display_name,
+        inputType: this.customAttributeInputType(attr.attribute_display_type),
+        filterOperators: this.getOperatorTypes(attr.attribute_display_type),
+      };
+    });
+    AUTOMATIONS.message_created.conditions.push(...customAttributeTypes);
+    AUTOMATIONS.conversation_created.conditions.push(...customAttributeTypes);
+    AUTOMATIONS.conversation_updated.conditions.push(...customAttributeTypes);
```
Comment (pranavrajs) [inline app/javascript/dashboard/routes/dashboard/settings/automation/AddAutomationRule.vue:275]: @fayazara These logic seems to be repeated in the edit AutomationRule as well. Can we create a mixin or use same component in both places?

**Comment by tejaswinichile** [comment]
@fayazara I have tested it, and everything works fine.
Just one suggestion and we can take it in the later ticket, what/how should we display if the custom attributes in contacts and conversations have the same name? 

custom-attr

**Comment by fayazara** [comment]
> @fayazara I have tested it, and everything works fine. Just one suggestion and we can take it in the later ticket, what/how should we display if the custom attributes in contacts and conversations have the same name?
> 
> custom-attr

@pranavrajs Should I append - Conversation/Contact beside each label?

**Comment by pranavrajs** [comment]
@fayazara Please fix the code climate issues

File: app/javascript/shared/mixins/automations/methodsMixin.js [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+    name: 'Incoming Message',
+  },
+  {
+    id: 'outgoing',
+    name: 'Outgoing Message',
+  },
+];
+export default {
+  methods: {
+    customAttributeInputType(key) {
```
Comment (pranavrajs) [inline app/javascript/shared/mixins/automations/methodsMixin.js:19]: Can't this be util function? It doesn't need to be a component method.

File: app/javascript/dashboard/mixins/automations/methodsMixin.js [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+      this.$v.$touch();
+      if (this.$v.$invalid) return;
+      const automation = JSON.parse(JSON.stringify(this.automation));
+      automation.conditions[
+        automation.conditions.length - 1
+      ].query_operator = null;
+      automation.conditions = filterQueryGenerator(
+        automation.conditions
+      ).payload;
+      automation.actions = actionQueryGenerator(automation.actions);
```
Comment (pranavrajs) [inline app/javascript/dashboard/mixins/automations/methodsMixin.js:163]: This can be a helper method, return a new automation object.

Comment (fayazara) [reply app/javascript/dashboard/mixins/automations/methodsMixin.js:163]: Yes, I have made some more changes to this, pushing the code in a bit @pranavrajs

File: app/javascript/dashboard/mixins/automations/methodsMixin.js [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+      ].conditions.find(
+        condition => condition.key === currentCondition.attribute_key
+      ).filterOperators[0].value;
+      this.automation.conditions[index].values = '';
+    },
+    showUserInput(operatorType) {
+      if (operatorType === 'is_present' || operatorType === 'is_not_present')
+        return false;
+      return true;
+    },
```
Comment (pranavrajs) [inline app/javascript/dashboard/mixins/automations/methodsMixin.js:182]:
```suggestion
    showUserInput(operatorType) {
      return operatorType != 'is_present' && operatorType != 'is_not_present';
    },
```

File: app/javascript/dashboard/mixins/automations/methodsMixin.js [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+    showActionInput(actionName) {
+      if (actionName === 'send_email_to_team' || actionName === 'send_message')
+        return false;
+      const type = this.automationActionTypes.find(
+        action => action.key === actionName
+      ).inputType;
+      if (type === null) return false;
+      return true;
```
Comment (pranavrajs) [inline app/javascript/dashboard/mixins/automations/methodsMixin.js:190]: Can't we combine this and `return !!type`

**Review by tejaswinichile — DISMISSED** [review]
LGTM 💯

File: app/javascript/dashboard/components/widgets/FilterInput/Index.vue [hunk 6]
```diff
@@ -215,6 +220,32 @@ export default {
         this.$emit('input', { ...payload, query_operator: value });
       },
     },
+    custom_attribute_type: {
```
Comment (nithindavid) [inline app/javascript/dashboard/components/widgets/FilterInput/Index.vue:223]: Is it possible to follow camelCase for computed properties?

**Comment by pranavrajs** [comment]
@fayazara A couple of issues. 

1. Disabled option for contact/conversation custom attributes is available in the dropdown even if there are no custom attributes available in the account.
2. The automations are not working.
```
Performing EventDispatcherJob (Job ID: 6d976f1f-4303-417c-925d-fdb031701fe6) from Sidekiq(events) enqueued at 2022-06-08T09:12:59Z with arguments: "conversation.updated", Wed, 08 Jun 2022 09:12:59.988382000 UTC +00:00, {:conversation=>#<GlobalID:0x00000001265472a8 @uri=#<URI::GID gid://chatwoot/Conversation/128222>>, :notifiable_assignee_change=>false, :changed_attributes=>{"updated_at"=>[Wed, 08 Jun 2022 09:12:58.276161000 UTC +00:00, Wed, 08 Jun 2022 09:12:59.983224000 UTC +00:00], "custom_attributes"=>[{"priority_1"=>"P1"}, {"priority_1"=>"P0"}]}, :performed_by=>nil}
Error performing EventDispatcherJob (Job ID: 6d976f1f-4303-417c-925d-fdb031701fe6) from Sidekiq(events) in 83.95ms: ActiveRecord::StatementInvalid (PG::SyntaxError: ERROR:  syntax error at or near ")"
LINE 1: ...rsations.id WHERE "conversations"."id" = $1 AND ( ) LIMIT $2
                                                             ^
):
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/connection_adapters/postgresql_adapter.rb:672:in `exec_params'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/connection_adapters/postgresql_adapter.rb:672:in `block (2 levels) in exec_no_cache'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activesupport-6.1.5.1/lib/active_support/dependencies/interlock.rb:48:in `block in permit_concurrent_loads'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activesupport-6.1.5.1/lib/active_support/concurrency/share_lock.rb:187:in `yield_shares'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activesupport-6.1.5.1/lib/active_support/dependencies/interlock.rb:47:in `permit_concurrent_loads'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/connection_adapters/postgresql_adapter.rb:671:in `block in exec_no_cache'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/connection_adapters/abstract_adapter.rb:696:in `block (2 levels) in log'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activesupport-6.1.5.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:26:in `block (2 levels) in synchronize'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activesupport-6.1.5.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `handle_interrupt'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activesupport-6.1.5.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `block in synchronize'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activesupport-6.1.5.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `handle_interrupt'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activesupport-6.1.5.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `synchronize'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/connection_adapters/abstract_adapter.rb:695:in `block in log'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activesupport-6.1.5.1/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/connection_adapters/abstract_adapter.rb:687:in `log'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/connection_adapters/postgresql_adapter.rb:670:in `exec_no_cache'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/connection_adapters/postgresql_adapter.rb:649:in `execute_and_clear'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/connection_adapters/postgresql/database_statements.rb:55:in `exec_query'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/connection_adapters/abstract/database_statements.rb:532:in `select'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/connection_adapters/abstract/database_statements.rb:69:in `select_all'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/connection_adapters/abstract/query_cache.rb:101:in `block in select_all'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/connection_adapters/abstract/query_cache.rb:118:in `block in cache_sql'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activesupport-6.1.5.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:26:in `block (2 levels) in synchronize'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activesupport-6.1.5.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `handle_interrupt'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activesupport-6.1.5.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `block in synchronize'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activesupport-6.1.5.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `handle_interrupt'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activesupport-6.1.5.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `synchronize'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/connection_adapters/abstract/query_cache.rb:109:in `cache_sql'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/connection_adapters/abstract/query_cache.rb:101:in `select_all'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/connection_adapters/abstract/database_statements.rb:95:in `select_rows'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/relation/finder_methods.rb:320:in `block in exists?'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/relation.rb:861:in `skip_query_cache_if_necessary'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/relation/finder_methods.rb:320:in `exists?'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/relation.rb:266:in `empty?'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/relation.rb:278:in `any?'
/Users/pranavrajs/chatwoot/chatwoot/app/services/automation_rules/conditions_filter_service.rb:32:in `perform'
/Users/pranavrajs/chatwoot/chatwoot/app/listeners/automation_rule_listener.rb:12:in `block in conversation_updated'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/relation/delegation.rb:88:in `each'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activerecord-6.1.5.1/lib/active_record/relation/delegation.rb:88:in `each'
/Users/pranavrajs/chatwoot/chatwoot/app/listeners/automation_rule_listener.rb:11:in `conversation_updated'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/wisper-2.0.0/lib/wisper/broadcasters/send_broadcaster.rb:5:in `public_send'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/wisper-2.0.0/lib/wisper/broadcasters/send_broadcaster.rb:5:in `broadcast'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/wisper-2.0.0/lib/wisper/registration/object.rb:18:in `broadcast'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/wisper-2.0.0/lib/wisper/publisher.rb:43:in `block in broadcast'
/Users/pranavrajs/.rvm/rubies/ruby-3.0.4/lib/ruby/3.0.0/set.rb:344:in `each_key'
/Users/pranavrajs/.rvm/rubies/ruby-3.0.4/lib/ruby/3.0.0/set.rb:344:in `each'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/wisper-2.0.0/lib/wisper/publisher.rb:42:in `broadcast'
/Users/pranavrajs/chatwoot/chatwoot/app/dispatchers/async_dispatcher.rb:8:in `publish_event'
/Users/pranavrajs/chatwoot/chatwoot/app/jobs/event_dispatcher_job.rb:5:in `perform'
/Users/pranavrajs/.rvm/gems/ruby-3.0.4/gems/activejob-6.1.5.1/lib/active_job/execution.rb:48:in `block in perform_now'
```

**Comment by tejaswinichile** [comment]
- [ ] Custom attribute and case insensitive filtering issue

**Comment by skyblue** [comment]
This is very needed, It must be useful .

**Review by tejaswinichile — APPROVED** [review]
💯 Tested on local, working well for me.

**Review by tejaswinichile — DISMISSED** [review]
💯

File: app/javascript/shared/mixins/specs/automationMixin.spec.js [hunk 7]
```diff
@@ -0,0 +1,480 @@
...
+
+  //   const generatedPayload = {
+  //     name: 'Fayaz',
+  //     description: 'Hello',
+  //     event_name: 'conversation_created',
+  //     conditions: [
+  //       {
+  //         attribute_key: 'status',
+  //         filter_operator: 'equal_to',
+  //         values: ['open'],
```
Comment (pranavrajs) [inline app/javascript/shared/mixins/specs/automationMixin.spec.js:262]:
```suggestion
```

File: app/javascript/shared/mixins/specs/automationMixin.spec.js [hunk 8]
```diff
@@ -0,0 +1,464 @@
...
+  //       },
+  //     ],
+  //   };
+  //   const wrapper = createComponent([methodsMixin, validationsMixin], data);
+  //   wrapper.vm.submitAutomation();
+  //   await wrapper.vm.$nextTick();
+
+  //   expect(wrapper.emitted()).toEqual(generatedPayload);
+  // });
+
```
Comment (pranavrajs) [inline app/javascript/shared/mixins/specs/automationMixin.spec.js:262]:
```suggestion
```

**Comment by pranavrajs** [comment]
@tejaswinichile @fayazara The automation doesn't seem to be working properly. Can you please check this behavior? https://www.loom.com/share/04a032784cdb4c8ebb1137721953c35d

**Review by iamsivin — COMMENTED** [review]
@fayazara I tested it locally and found some issues.
https://www.loom.com/share/693bdf9fb2ce4d859f10b20121cb8493

1. Automation is not working as expected with the custom attribute condition check and without. It works sometimes. You can see it in the loom video.
2. If we saved automation with an action for a condition. If we change that action to another one, the action dropdown values are not getting cleared. You can see it in the loom video with "Assign label and Assign Team".
3. Now, we can select multiple teams. I think we can only assign one team at a time.

**Comment by fayazara** [comment]
> @fayazara I tested it locally and found some issues. https://www.loom.com/share/693bdf9fb2ce4d859f10b20121cb8493
> 
> 1. Automation is not working as expected with the custom attribute condition check and without. It works sometimes. You can see it in the loom video.
> 2. If we saved automation with an action for a condition. If we change that action to another one, the action dropdown values are not getting cleared. You can see it in the loom video with "Assign label and Assign Team".
> 3. Now, we can select multiple teams. I think we can only assign one team at a time.

@iamsivin Fixed 2 and 3.

@tejaswinichile Can you please check 1, I was able to reproduce this, below is the automation.

Payload
```js
{
  "account_id": 1,
  "name": "Check",
  "description": "Hello",
  "event_name": "message_created",
  "conditions": [
    {
      "values": [
        "outgoing"
      ],
      "attribute_key": "message_type",
      "query_operator": "and",
      "filter_operator": "equal_to",
      "custom_attribute_type": ""
    },
    {
      "values": [
        false
      ],
      "attribute_key": "vfer",
      "filter_operator": "equal_to",
      "custom_attribute_type": "contact_attribute"
    }
  ],
  "actions": [
    {
      "action_name": "add_label",
      "action_params": [
        "sales",
        "billing"
      ]
    }
  ]
}
```

**Review by tejaswinichile — APPROVED** [review]
LGTM 💯

**Comment by pranavrajs** [comment]
@iamsivin @muhsin-k Can you please test this PR? Please sync up with @fayazara if needed.

**Review by iamsivin — COMMENTED** [review]
@fayazara @tejaswinichile I have done QA locally. 
I faced some issues.
https://www.loom.com/share/266c8e3c75c145e4b6920d04135566c1

1. Assign team action is not working. Value is not saving.
2. Assign label is working, but it is showing multiple times.
3. I Faced some rendering issues with the label in the sidebar. When I click the clear button, it will remove all values from the store, but you can see it back after refreshing.
4. Show some console warnings from i18n.
5. Cloning an existing automation which has an attachment uploaded. The newly created one will not have an attachment. You can see it in loom video https://www.loom.com/share/f0261bf61c8f4550b52bbac7a493e293

You can see it all in the loom video.

**Comment by iamsivin** [comment]
> @iamsivin @muhsin-k Can you please test this PR? Please sync up with @fayazara if needed.

Done https://github.com/chatwoot/chatwoot/pull/4548#pullrequestreview-1167364495

**Review by iamsivin — COMMENTED** [review]
@fayazara @tejaswinichile I faced another issue while doing QA. The automation is getting triggered multiple times. You can see it in the loom video.
https://www.loom.com/share/af9901ff8dc2401f801571fbcd61492d

**Comment by fayazara** [comment]
> @fayazara @tejaswinichile I faced another issue while doing QA. The automation is getting triggered multiple times. You can see it in the loom video. https://www.loom.com/share/af9901ff8dc2401f801571fbcd61492d



> @fayazara @tejaswinichile I faced another issue while doing QA. The automation is getting triggered multiple times. You can see it in the loom video. https://www.loom.com/share/af9901ff8dc2401f801571fbcd61492d

@tejaswinichile Was able to reproduce this, and it happened for this specific automation.  Please have a look

Event - `message created`
Condition -`conversation attribute equals`
Actions- 
```
Assign Team
Assign Label
Send Attachment
Send Message
```
I tried other automations like -
1. Event - `Message created`, `Custom Attribute link equals, `Assign Label`  - worked ✅
2. Event - `Conversation Updatesd` , `Status equals Resolved`, `Assign Team` - worked ✅

**Comment by muhsin-k** [comment]
> Event - `message created` Condition -`conversation attribute equals` Actions-
> 
> ```
> Assign Team
> Assign Label
> Send Attachment
> Send Message
> ```

@fayazara @tejaswinichile 

I have tested this case in the staging environment. But the team assignment activity message is not coming.

https://www.loom.com/share/574500075dde40c3a4a43bda77a804a2

**Comment by tejaswinichile** [comment]
@muhsin-k @fayazara I will resolve this as a separate bug ticket. Please move ahead with this PR merge. It's been pending for a long time. 

I will handle it in this sprint iteration only.

File: app/javascript/dashboard/mixins/automations/methodsMixin.js [hunk 9]
```diff
@@ -0,0 +1,289 @@
+import languages from '../../../dashboard/components/widgets/conversation/advancedFilterItems/languages.js';
+import countries from '../../../shared/constants/countries.js';
```
Comment (iamsivin) [inline app/javascript/dashboard/mixins/automations/methodsMixin.js:2]:
```suggestion
import languages from 'dashboard/components/widgets/conversation/advancedFilterItems/languages';
import countries from 'shared/constants/countries';
```

File: app/javascript/shared/mixins/specs/automationHelper.spec.js [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+} from '../../../dashboard/routes/dashboard/settings/automation/operators';
+import {
+  customAttributes,
+  labels,
+  automation,
+  contactAttrs,
+  conversationAttrs,
+  expectedOutputForCustomAttributeGenerator,
+} from './automationFixtures';
+import { AUTOMATIONS } from '../../../dashboard/routes/dashboard/settings/automation/constants.js';
```
Comment (iamsivin) [inline app/javascript/shared/mixins/specs/automationHelper.spec.js:15]:
```suggestion
import * as helpers from 'dashboard/helper/automationHelper';
import {
  OPERATOR_TYPES_1,
  OPERATOR_TYPES_3,
  OPERATOR_TYPES_4,
} from 'dashboard/routes/dashboard/settings/automation/operators';
import {
  customAttributes,
  labels,
  automation,
  contactAttrs,
  conversationAttrs,
  expectedOutputForCustomAttributeGenerator,
} from './automationFixtures';
import { AUTOMATIONS } from 'dashboard/routes/dashboard/settings/automation/constants';
```

File: app/javascript/dashboard/mixins/automations/methodsMixin.js [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+  isACustomAttribute,
+  getFileName,
+  getDefaultConditions,
+  getDefaultActions,
+  filterCustomAttributes,
+  generateAutomationPayload,
+  getStandardAttributeInputType,
+  isCustomAttribute,
+  generateCustomAttributes,
+} from '../../helper/automationHelper.js';
```
Comment (iamsivin) [inline app/javascript/dashboard/mixins/automations/methodsMixin.js:18]:
```suggestion
} from 'dashboard/helper/automationHelper';
```

File: app/javascript/dashboard/helper/automationHelper.js [hunk 12]
```diff
@@ -0,0 +1,242 @@
+import {
+  OPERATOR_TYPES_1,
+  OPERATOR_TYPES_3,
+  OPERATOR_TYPES_4,
+} from '../routes/dashboard/settings/automation/operators';
+import filterQueryGenerator from './filterQueryGenerator.js';
+import actionQueryGenerator from './actionQueryGenerator.js';
```
Comment (iamsivin) [inline app/javascript/dashboard/helper/automationHelper.js:7]:
```suggestion
import {
  OPERATOR_TYPES_1,
  OPERATOR_TYPES_3,
  OPERATOR_TYPES_4,
} from 'dashboard/routes/dashboard/settings/automation/operators';
import filterQueryGenerator from './filterQueryGenerator';
import actionQueryGenerator from './actionQueryGenerator';
```

**Review by iamsivin — COMMENTED** [review]
LGTM 👍🏻 @fayazara Please see the comments added.

**Comment by fayazara** [comment]
> LGTM 👍🏻 @fayazara Please see the comments added.

@iamsivin Updated

**Review by muhsin-k — APPROVED** [review]

**Review by iamsivin — APPROVED** [review]

**Review by pranavrajs — APPROVED** [review]

**Pull request merged**